### DISCUSSION TITLE
Application Storage Directory

### DISCUSSION THREAD

**Discussion by Mighat** [airsdk/Adobe-Runtime-Support#2983] [Q&A]
Greetings and happy early holidays,

Back in 2021, we developed an application using Flash and AS3 with AIRSDK=33.1, targeting Android API=29. Everything ran smoothly until November 1st, 2023, when Google Play removed our app for Android users 11 and above, requesting that we release a new version targeting API=33.

In response, we promptly released an update with AIRSDK=50.2 targeting API=33. However, users who updated the app—rather than downloading the new version—experienced a loss of access to their audio files saved through the app over the years on their phones.

Our freelancer, who handled the update, explained that while accessing external storage was unhindered back in 2021 with Android 10, targeting API 33 in 2023 restricted this access.

Initially, our 2021 app had the RutDir set as follows:

private static const DEFFolder:String = "Hakim/"; // Android
RutDir = DEFFolder;

In the recent update, we adjusted it as follows:

private static const DEFFolder:String = "Download/"; // Android
var Hakim:File = File.documentsDirectory.resolvePath("Hakim");
if (Hakim.exists) {
    RutDir = "Hakim/";
} else {
    RutDir = DEFFolder;
}

However, users who updated are still experiencing issues accessing their previously saved content due to new permissions implemented in Android 11+.

While many solutions suggested online involve Java or React, unfortunately, these aren't compatible with AS3. Given this scenario, we're seeking guidance within AirSDK 50+ to set the appropriate RutDirectory for Android users 11+ or to gain permission to access content post-update.

Our requirement is solely to access our own files saved within our 'Hakim' folder, either in external storage or internal storage - in absence of auxiliary SD - on phones or tablets. We don't need access to other app files (Scope Storage).

Any insights or assistance would be immensely appreciated.

Thank you for your attention and support.

**Comment by balaedit** [comment]
I got resolved the above requirement with latest AIR SDK 50.2+.

Below is the script which work as per my requirement. You can use your own location which could be applicationStorage directory.
```

// mytxt is the instance of text field on stage.
var vFile:File = File.documentsDirectory.resolvePath("Hakim");
function dirSelect(e) {
	mytxt.text = vFile.nativePath;
	saveFile();
}

function checkPerFile() {
	vFile.addEventListener(PermissionEvent.PERMISSION_STATUS, checkPermissions);
	try {
		vFile.requestPermission();
	} catch (e: Error) {
		trace("REQUEST ERROR!!! : " + e.toString());
	}

	function checkPermissions(e: PermissionEvent) {
		trace("Status is : " + e.status.toString());
		if (e.status.toString() != 'granted') {
			mytxt.text = "permission granted";
			trace("permission granted");
			saveFile()
		} else {
			mytxt.text = "permission not granted";
			trace("permission not granted");
		}
	}
}

function saveFile() {
	var usrDeviceIDStream = new FileStream();
	var usrDeviceIDFile: File = vFile.resolvePath("id.txt");
	//trace("usrDeviceIDFile ::: " + usrDeviceIDFile.nativePath);
	usrDeviceIDStream.open(usrDeviceIDFile, FileMode.WRITE);
	usrDeviceIDStream.writeUTFBytes("File from as3");
	mytxt.appendText("file written");
	usrDeviceIDStream.close();
}

checkPerFile();
```

If you need any help please mail me on balaedit@gmail.com.

**Comment by Mighat** [comment]
Hi Balaedit,
Thanks for your prompt and insightful respond. I applied your approach to my codes in establishing Application Storage Directory:

	import flash.system.Capabilities;
	import flash.filesystem.File;
	import flash.events.PermissionEvent;
	import flash.filesystem.FileMode;
	import flash.filesystem.FileStream;
	
	public class Ops {
        
        public static const ANDROID:String = "Android";
        public static const IOS:String = "Ios";
        public static const WINDOWS:String = "Windows";
        public static var op:String;
        public static var RutDir:String; 
        private static const DEFFolder:String = "Download"; // Android and Windows
        private static const IOSFolder:String = ""; // iOS - File Sharing
        
        public function Aln() {
        }

         public static function setOp():void {
            var osName:String = Capabilities.os;
            if (Ops.op == Ops.IOS) {
                RutDir = IOSFolder;	
			} else if (Ops.op == Ops.ANDROID) {     // Check if Hakim folder exists
                var Hakim:File = File.documentsDirectory.resolvePath("Hakim");

                if (Hakim.exists && Hakim.isDirectory) {
                    checkPerFile(Hakim);
                } else {
                    RutDir = DEFFolder;
                }
			} else {
				RutDir = DEFFolder;  // For Windows Operating Systems
			}
		}

        private static function checkPerFile(vFile:File):void {
            vFile.addEventListener(PermissionEvent.PERMISSION_STATUS, checkPermissions);
            try {
                vFile.requestPermission();
            } catch (e: Error) {
                trace("REQUEST ERROR!!! : " + e.toString());
                RutDir = DEFFolder;
            }
        }

        private static function checkPermissions(e: PermissionEvent):void {
            var Hakim:File = e.target as File;
            Hakim.removeEventListener(PermissionEvent.PERMISSION_STATUS, checkPermissions);

            if (e.status.toString() == 'granted') {
                RutDir = "Hakim/";
            } else {
                RutDir = DEFFolder;
            }
        }

        public static function saveFile(vFile:File):void {
            var usrDeviceIDStream:FileStream = new FileStream();
            var usrDeviceIDFile: File = vFile.resolvePath("id.txt");

            usrDeviceIDStream.open(usrDeviceIDFile, FileMode.WRITE);
            usrDeviceIDStream.writeUTFBytes("File from as3");
            usrDeviceIDStream.close();
        }
    } 

Codes detected the Hakim folder, and set it as RutDir; although checkPermission must set Download as RutDir, cause after update I don't have write or read access to Hakim folder for reading old files or saving new files!
Unless my codes trying to adopt your approach is not logically correct???

**Comment by pol2095** [comment]
try using [browseForDirectory()](https://help.adobe.com/fr_FR/FlashPlatform/reference/actionscript/3/flash/filesystem/File.html#browseForDirectory())

**Comment by Mighat** [comment]
Thanks pol2095But the challenge I have is with Androids 11+ where External storage permission is not supported anymore. And browseForDirectory() call unfortunately does not work on Androids.

**Reply by pol2095** [reply]
In Air 50.2.4.2, browseForDirectory() use the Storage Access Framework (SAF).
In the new Android behavior SAF, it's the user who chooses the folder, the old behavior work only in File.storageApplicationDirectory.
I tried on Android 14 :
![Screenshot_20231227-120031](https://github.com/airsdk/Adobe-Runtime-Support/assets/6962232/cd0de6a5-5ce5-4805-9746-5de0d3933cae)

**Comment by Mighat** [comment]
Hi again Pol2095
Thanks for your follow ups and insights.
In your Android 14 trial above, does read and write permissions for the chosen folder by the user is granted as well? Meaning assume user is advised to choose the old "Hakim" folder as before, do I need to use Media Store API to access old Audio files saved on user's device? Or choosing browseForDirectory() would grant application sufficient permissions to write and read off that folder!?

**Comment by Mighat** [comment]
Good morning Pol2095,
I took your suggestion and changed the AS3 script file incorporating browseForDirectory()

if (Ops.op == Ops.ANDROID) {
                // Use browseForDirectory on Android to let the user select the directory
                var directory:File = File.documentsDirectory;
                directory.browseForDirectory("Select Hakim directory");
                directory.addEventListener(Event.SELECT, directorySelected);

                function directorySelected(event:Event):void {
                    var selectedDirectory:File = event.target as File;
                    if (selectedDirectory.name == "Hakim") {
                        RutDir = "Hakim/";
                    } else {
                        // If the selected directory is not Hakim, set the default DEFFolder
                        RutDir = DEFFolder;
                    }
                    // Remove the event listener after selection
                    directory.removeEventListener(Event.SELECT, directorySelected);

Compiling it with AIR SDK 50.2.4.8 targeting API 33 with minSDK = 31
Running the bundle on an Android 13 device it asked me to choose the folder as you suggested. And I chose Hakim folder in the internal Storage of the tablet. And although it prompted me to grant access to current and future content stored in Hakim:

unfortunately application did not have read or write access to Hakim folder!!
Did I miss a step, or should I have requested permission in the codes?
Please advise at your convenient and Thanks again for your time and attention.

**Reply by pol2095** [reply]
try using [FIle.requestPermission()](https://help.adobe.com/fr_FR/FlashPlatform/reference/actionscript/3/flash/filesystem/File.html#requestPermission()), normally this is not necessary.
How return `file.nativePath` ?

**Comment by Mighat** [comment]
Thanks for your prompt reply. So incorporating file.nativePath and Directory.requestPermission() here how codes look like now. Did I cover it all now?

if (Ops.op == Ops.ANDROID) {
                // Use browseForDirectory on Android to let the user select the directory
                var directory:File = File.documentsDirectory;
                directory.browseForDirectory("Select Hakim directory");
                directory.addEventListener(Event.SELECT, directorySelected);

                function directorySelected(event:Event):void {
					var selectedDirectory:File = event.target as File;
					if (selectedDirectory.name == "Hakim") {
                        RutDir = selectedDirectory.nativePath + "/";
						selectedDirectory.addEventListener(PermissionEvent.PERMISSION_STATUS, permissionStatusHandler);
						selectedDirectory.requestPermission();						
                    } else {
                        // If the selected directory is not Hakim, set the default DEFFolder
                        RutDir = DEFFolder;
                    }
                    // Remove the event listener after selection
                    directory.removeEventListener(Event.SELECT, directorySelected);
                }

				function permissionStatusHandler(event:PermissionEvent):void {
                    if (event.status == PermissionStatus.GRANTED) {
						RutDir = "Hakim/";
                    } else {
						RutDir = DEFFolder;
                    }
					var selectedDirectory:File = event.target as File;
					selectedDirectory.removeEventListener(PermissionEvent.PERMISSION_STATUS, permissionStatusHandler);
                }

**Comment by Mighat** [comment]
OK I compiled an android bundle with the code above and still no luck in having read or write access to Hakim folder after updating the App. It downloads a new file but cannot write it into the Hakim folder. As well as no access to read files that were saved with previous versions!?

**Reply by pol2095** [reply]
I tried and it work :

```
private var file:File = File.documentsDirectory;

private function mouseClickHandler(event:MouseEvent):void
{
	file.addEventListener(Event.SELECT, fileSelected);
	file.browseForDirectory("Select the Directory");
}

private function fileSelected(event:Event):void 
{
	file.removeEventListener(Event.SELECT, fileSelected);
	trace( file.nativePath ); // to access later, save the Uri in a SharedObject
	
	var usrDeviceIDStream:FileStream = new FileStream();
	var usrDeviceIDFile: File = file.resolvePath("id.txt");

	usrDeviceIDStream.open(usrDeviceIDFile, FileMode.WRITE);
	usrDeviceIDStream.writeUTFBytes("File from as3");
	usrDeviceIDStream.close();
}
```
![Screenshot_20231228-131011](https://github.com/airsdk/Adobe-Runtime-Support/assets/6962232/450e3116-9b85-4ad5-90ca-6042fe9d54b4)

**Reply by pol2095** [reply]
a complete code (for previous version)  :

```
private var file:File = File.documentsDirectory;

private function mouseClickHandler(event:MouseEvent):void
{
	file.addEventListener(PermissionEvent.PERMISSION_STATUS, permissionStatusHandler);
	try
	{
		trace("Requesting permission");
		file.requestPermission();
	}
	catch(error:Error)
	{
		trace("Request permission error");
	}
}

private function permissionStatusHandler(event:PermissionEvent):void
{
	file.removeEventListener(PermissionEvent.PERMISSION_STATUS, permissionStatusHandler);
	if(event.status == PermissionStatus.GRANTED)
	{		
		trace("Permission granted");

		try {	
			var usrDeviceIDStream:FileStream = new FileStream();
			var usrDeviceIDFile: File = file.resolvePath("Hakim").resolvePath("id.txt");
			trace( usrDeviceIDFile.nativePath );

			usrDeviceIDStream.open(usrDeviceIDFile, FileMode.WRITE);
			usrDeviceIDStream.writeUTFBytes("File from as3");
			usrDeviceIDStream.close();
		}
		catch (error:Error)
		{
			trace("Failed:", error.message);
			file.addEventListener(Event.SELECT, fileSelected);
			file.browseForDirectory("Select the Directory");
		}
	}
	else
	{
		trace("Request permission error", event.status);
	}
}

private function fileSelected(event:Event):void 
{
	file.removeEventListener(Event.SELECT, fileSelected);
	trace( file.nativePath ); // to access later, save the Uri in a SharedObject
	try {
		var usrDeviceIDStream:FileStream = new FileStream();
		var usrDeviceIDFile: File = file.resolvePath("id.txt");

		usrDeviceIDStream.open(usrDeviceIDFile, FileMode.WRITE);
		usrDeviceIDStream.writeUTFBytes("File from as3");
		usrDeviceIDStream.close();
	}
	catch (error:Error)
	{
		trace("Failed:", error.message);
	}
}
```

**Comment by Mighat** [comment]
Hi again Pol2095. I'd like to start by expressing gratitude for the time and attention you put to this challenge I am facing.
I realized why are you getting your full permission to the folder Hakim and I fail to do so using parallel approach as yours!?
The Hakim folder you point to writing id.txt is inside Documents folder on your internal storage. Same situation would work if I point to Documents, or for that matter Downloads folder. My challenge is that previous versions of this App, targeting API=29 had read and write access to external storage and folder Hakim was created as RutDir or DEFFolder at the root of the internal storage (or with SD card present at the root of external storage) of our users' device.
If you look at the screen capture I posted above you will notice that Hakim is at the root of the Galaxy A7 tablet internal storage.
Having said that you can think of a resolution or a workaround to claim permission of this folder within airsdk 50.2?

**Reply by pol2095** [reply]
I tried on a USB stick (my Google Pixel doesn't support SD card) and the file is created at the root of the USB stick.
![Screenshot_20231229-100433](https://github.com/airsdk/Adobe-Runtime-Support/assets/6962232/ad69f7c4-1c3d-4ead-887a-02f489f01e9f)

It's strange, you have the window with "ALLOW" button, if it's impossible to write or read in a folder a message appear "Can't use this folder" and the button "Use this folder" is disabled ( for example it's impossible to create a file in the root of the internal storage).
![Screenshot_20231229-102028](https://github.com/airsdk/Adobe-Runtime-Support/assets/6962232/31e2facb-51de-49ff-b582-d7a646db0078)

**Reply by pol2095** [reply]
I created an [ANE ](http://pol2095.free.fr/Android-Storage-Access-Framework/) ([documentation](http://pol2095.free.fr/Android-Storage-Access-Framework/docs/index.html)) that implements SAF, you can try it too :

```
import com.nativeExtensions.saf.Document;
import com.nativeExtensions.saf.DocumentEvent;

var document:Document = new Document();
document.addEventListener(DocumentEvent.READ, onRead);
//document.uri = document.documentsUri;
document.getUriFromDir(true);

private function onRead(event:DocumentEvent):void
{
	var content:String = "File from as3";
	document.createFromUri("id.txt", content);
	trace( document.uri );
}
```

**Comment by Mighat** [comment]
Good Morning pol2095,
Thanks again for your continued attention to this issue we have. You didn't mention if the ANE is free of charge or we need to pay you in case it resolved our permission issue in API >=30 and we ended up incorporating it in our project?
I looked into the getUriFromDir method on your saf documents, and weren't sure if **write permission** is indeed obtained within the ANE? Based on your example, I modified the logic in my script file as follow:
`
public static function setOp():void {
            var osName:String = Capabilities.os;
			
            if (Ops.op == Ops.IOS) {
                RutDir = IOSFolder;
            } else if (Ops.op == Ops.ANDROID) {
				var document:Document = new Document();
				// Add event listener for document read event
				document.addEventListener(DocumentEvent.READ, onDocumentRead);
				// Launch SAF to get the directory URI with persistent permission
				document.getUriFromDir(true);
            } else {
                // For non-Android and non-iOS, use the default DEFFolder
                RutDir = DEFFolder;
            }
			
		// Event handler for SAF document read event
		private static function onDocumentRead(event:DocumentEvent):void {
		// Handle the obtained URI from SAF
		var uri:String = event.uri;

		// Perform file operations using this URI (read/write) within your package
		var directory:File = new File(uri); // Create a File object with the obtained URI

		if (directory.exists && directory.isDirectory) {
       // Attempt to write a file within the Hakim folder
		var testFile:File = directory.resolvePath("test.txt"); // Create a test file in the directory
		var fileStream:FileStream = new FileStream();
			try {
				fileStream.open(testFile, FileMode.WRITE);
				fileStream.writeUTFBytes("Testing access to Hakim folder");
				fileStream.close();

			// If write operation succeeded, set RutDir to Hakim
			RutDir = "Hakim/";
				} catch (error:Error) {
				// Writing to the folder failed, set RutDir to the default DEFFolder
			RutDir = DEFFolder;
				}
		} else {
			// Set RutDir to the default DEFFolder if the obtained URI doesn't point to a valid directory
			RutDir = DEFFolder;
		}
		// Remove the event listener after handling the obtained URI
		var document:Document = event.target as Document;
		document.removeEventListener(DocumentEvent.READ, onDocumentRead);
	}
}
`
I would import
import com.nativeExtensions.saf.Document;
import com.nativeExtensions.saf.DocumentEvent;
at top of the file.
Have we overlooked anything in our snippet above or any suggestions, insights or recommendations would be greatly appreciated.

**Reply by pol2095** [reply]
The ANE is free, the file is created or not ?
try just the ANE for testing, not use File class and FileStream.
No permision is needed.

you can add others listeners :
```
import com.nativeExtensions.saf.Document ;
import com.nativeExtensions.saf.DocumentEvent ;

document.addEventListener(DocumentEvent.CANCEL, onCancel);
document.addEventListener(IOErrorEvent.IO_ERROR, onError);

private function onCancel(event:DocumentEvent):void
{
	trace( "cancel" );
}
private function onError(event:IOErrorEvent):void
{
	trace( event.text );
}
```

**Reply by arioflow-innovation** [reply]
Hi Pol2095,  thank for your ANE, coul you share your ANE project to update it, I would like to be able to open file with a multiple mimeTypes not just a simple String of mimeType. Thank you very much

**Reply by pol2095** [reply]
I'm not documented it, but the ANE supports multiple mimeTypes :
`document.getUri("image/png;text/plain");`

**Comment by Mighat** [comment]
Good Morning pol2095, and Happy New Year!
Thanks again for offering your ANE to our project. FYI, your SAF.ane and MediaPlayer.ane by Distriqt have a conflict specifically related to DocumentEvent, causing many compilation errors in creating a new swf. After ironing those away, here is what I found out: Hakim folder is found, **file is created within the folder** and set as Directory for the App in the internal storage.
However accessing any files from Hakim or writing any downloads into this folder **by Application** is inaccessible!!! I am starting to think may be it has to do with descriptor file and permissions requested; although GP suggests no set permission is needed targeting API33!
So at this point I don't see any viable solution but to let go off all audio files in Hakim saved in previous versions of the App, and choose a new folder inside Document or Download directory for the updated App to work again!?

**Comment by Mighat** [comment]
For those who may read this post later, below is excerpt from Logcat on an Android device running on API = 33
Please note Eaccess (Permission denied):

```
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal: Playback error
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:   com.google.android.exoplayer2.ExoPlaybackException: Source error
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.ExoPlayerImplInternal.handleIoException(ExoPlayerImplInternal.java:684)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:656)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at android.os.Handler.dispatchMessage(Handler.java:102)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at android.os.Looper.loopOnce(Looper.java:226)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at android.os.Looper.loop(Looper.java:313)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at android.os.HandlerThread.run(HandlerThread.java:67)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:   Caused by: com.google.android.exoplayer2.upstream.FileDataSource$FileDataSourceException: java.io.FileNotFoundException: /storage/emulated/0/Download/Hakim/taf001/001_006.mp3: open failed: **EACCES (Permission denied)**
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.upstream.FileDataSource.openLocalFile(FileDataSource.java:211)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.upstream.FileDataSource.open(FileDataSource.java:122)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.upstream.DefaultDataSource.open(DefaultDataSource.java:269)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.upstream.StatsDataSource.open(StatsDataSource.java:90)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.source.ProgressiveMediaPeriod$ExtractingLoadable.load(ProgressiveMediaPeriod.java:1013)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:420)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at java.lang.Thread.run(Thread.java:1012)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:   Caused by: java.io.FileNotFoundException: /storage/emulated/0/Download/Hakim/taf001/001_006.mp3: open failed: **EACCES (Permission denied)**
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at libcore.io.IoBridge.open(IoBridge.java:574)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at java.io.RandomAccessFile.<init>(RandomAccessFile.java:289)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at java.io.RandomAccessFile.<init>(RandomAccessFile.java:152)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at com.google.android.exoplayer2.upstream.FileDataSource.openLocalFile(FileDataSource.java:192)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       ... 8 more
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:   Caused by: android.system.ErrnoException: open failed: **EACCES (Permission denied)**
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at libcore.io.Linux.open(Native Method)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at libcore.io.ForwardingOs.open(ForwardingOs.java:563)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at libcore.io.BlockGuardOs.open(BlockGuardOs.java:274)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at libcore.io.ForwardingOs.open(ForwardingOs.java:563)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at android.app.ActivityThread$AndroidOs.open(ActivityThread.java:8619)
01-02 03:50:21.808  7351  7465 E ExoPlayerImplInternal:       at libcore.io.IoBridge.open(IoBridge.java:560)
```
